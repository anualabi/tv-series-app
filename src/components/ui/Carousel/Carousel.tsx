'use client'

import { motion, useAnimation } from 'framer-motion';
import React, { useRef, useState, useEffect } from 'react';

import { IconArrowLeft, IconArrowRight } from '@/components/icons';

export type CarouselProps = {
    items: React.ReactNode[];
    itemWidth?: number;
};

const ArrowButton: React.FC<{
  direction: 'left' | 'right';
  onClick: () => void;
  disabled: boolean;
}> = ({ direction, onClick, disabled }) => {
  const Icon = direction === 'left' ? IconArrowLeft : IconArrowRight;

  return (
    <motion.button
      whileHover={{ scale: 1.1 }}
      whileTap={{ scale: 0.9 }}
      className={`mx-2 text-white focus:outline-none ${disabled ? 'opacity-50 cursor-not-allowed' : ''}`}
      onClick={onClick}
      disabled={disabled}
      data-testid={direction === 'left' ? "left-arrow" : "right-arrow"}
    >
      <Icon />
    </motion.button>
  );
};

const Carousel: React.FC<CarouselProps> = ({ items, itemWidth = 250 }) => {
    const boxRef = useRef(null);
    const controls = useAnimation();
    const [position, setPosition] = useState<number>(0);
    const [shouldDrag, setShouldDrag] = useState<boolean>(false);

    useEffect(() => {
        const updateShouldDrag = () => {
            setShouldDrag(window.innerWidth <= 768);
        };
        window.addEventListener('resize', updateShouldDrag);

        updateShouldDrag();

        return () => window.removeEventListener('resize', updateShouldDrag);
    }, []);

    const handleNext = () => {
        if (position > -(items.length - 1) * itemWidth) {
            setPosition(position - itemWidth);
            controls.start({ x: position - itemWidth });
        }
    };

    const handlePrevious = () => {
        if (position < 0) {
        setPosition(position + itemWidth);
        controls.start({ x: position + itemWidth });
        }
    };

    return (
        <motion.div className="h-full flex w-full overflow-x-hidden relative" ref={boxRef}>
            <motion.div
                className="inline-flex gap-4"
                drag={shouldDrag ? 'x' : false}
                dragConstraints={boxRef}
                animate={controls}
                transition={{ ease: "easeOut", duration: 1 }}
            >
                {items}
            </motion.div>

            {/* Left and Righ arrow icons */}
            <div className="absolute bottom-0 right-3 hidden md:block">
                <ArrowButton direction="left" onClick={handlePrevious} disabled={position >= 0} />
                <ArrowButton direction="right" onClick={handleNext} disabled={position <= -(items.length - 1) * itemWidth / 2} />  
            </div>
        </motion.div>
    );
};

export default Carousel;
