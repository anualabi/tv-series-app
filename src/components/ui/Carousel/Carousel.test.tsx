import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Carousel, { CarouselProps} from './Carousel';

const items = [
    <div key="1">Item 1</div>,
    <div key="2">Item 2</div>,
    <div key="3">Item 3</div>
];

const mockProps = {
    items: items,
    itemWidth: 300,
};

const customRender = (props: CarouselProps) => render(<Carousel {...props} />);

describe('Carousel', () => {
    test('should render snapshot', () => {
        const { asFragment } = customRender({ ...mockProps });
        
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders items passed to it', () => {
        customRender({...mockProps})
        expect(screen.getByText('Item 1')).toBeInTheDocument();
        expect(screen.getByText('Item 2')).toBeInTheDocument();
    });

    test('handles next and previous actions', () => {
        customRender({...mockProps})

        const leftArrowButton = screen.getByTestId('left-arrow');
        const rightArrowButton = screen.getByTestId('right-arrow');

        expect(leftArrowButton).toBeDisabled();

        fireEvent.click(rightArrowButton);

        expect(leftArrowButton).toBeEnabled();

        fireEvent.click(leftArrowButton);

        expect(leftArrowButton).toBeDisabled();
    });

});