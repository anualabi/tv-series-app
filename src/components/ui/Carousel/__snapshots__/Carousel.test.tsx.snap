// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full flex w-full overflow-x-hidden relative"
  >
    <div
      class="inline-flex gap-4"
    >
      <div>
        Item 1
      </div>
      <div>
        Item 2
      </div>
      <div>
        Item 3
      </div>
    </div>
    <div
      class="absolute bottom-0 right-3 hidden md:block"
    >
      <button
        class="mx-2 text-white focus:outline-none opacity-50 cursor-not-allowed"
        data-testid="left-arrow"
        disabled=""
        tabindex="0"
      >
        <svg
          height="19.414"
          viewBox="0 0 28.707 19.414"
          width="28.707"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="tail-left"
            transform="translate(29.207 19.207) rotate(-180)"
          >
            <path
              d="M0,0H28"
              fill="none"
              id="Shape"
              stroke="#fff"
              stroke-miterlimit="10"
              stroke-width="1"
              transform="translate(0.5 9.5)"
            />
            <path
              d="M0,0,9,9,0,18"
              data-name="Shape"
              fill="none"
              id="Shape-2"
              stroke="#fff"
              stroke-linecap="square"
              stroke-miterlimit="10"
              stroke-width="1"
              transform="translate(19.5 0.5)"
            />
          </g>
        </svg>
      </button>
      <button
        class="mx-2 text-white focus:outline-none "
        data-testid="right-arrow"
        tabindex="0"
      >
        <svg
          height="19.414"
          viewBox="0 0 28.707 19.414"
          width="28.707"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="tail-right"
            transform="translate(-0.5 0.207)"
          >
            <path
              d="M0,0H28"
              fill="none"
              id="Shape"
              stroke="#fff"
              stroke-miterlimit="10"
              stroke-width="1"
              transform="translate(0.5 9.5)"
            />
            <path
              d="M0,0,9,9,0,18"
              data-name="Shape"
              fill="none"
              id="Shape-2"
              stroke="#fff"
              stroke-linecap="square"
              stroke-miterlimit="10"
              stroke-width="1"
              transform="translate(19.5 0.5)"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
